server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepository.java

```diff
@@ -1292,109 +1262,112 @@ static void doRefUpdate(org.eclipse.jgit.lib.Repository jGitRepository, RevWalk
-                    });
-        } finally {
-            readUnlock();
-        }
+                if (filter.matches(path)) {
+                    return range.to();
+                }
+            }
 
-        return matches ? range.to() : null;
+            return null;
+        });
     }
 
     @Override
     public CompletableFuture<Revision> watch(Revision lastKnownRevision, String pathPattern) {
-        final ServiceRequestContext ctx =
-                RequestContext.mapCurrent(ServiceRequestContext.class::cast, null);
-        final CompletableFuture<Revision> future = new CompletableFuture<>();
-        CompletableFuture.runAsync(() -> {
-            if (ctx != null && ctx.isTimedOut()) {
-                logger.info("{} Ignoring watch operation. The request has already timed out: " +
-                            "project={}, repo={}, lastKnownRevision={}, pathPattern={}",
-                            ctx, parent.name(), name, lastKnownRevision, pathPattern);
-                throw REQUEST_ALREADY_TIMED_OUT;
-            }
+        requireNonNull(lastKnownRevision, "lastKnownRevision");
+        requireNonNull(pathPattern, "pathPattern");
 
-            requireNonNull(lastKnownRevision, "lastKnownRevision");
-            requireNonNull(pathPattern, "pathPattern");
+        final ServiceRequestContext ctx = context();
+        final Revision normLastKnownRevision = normalizeNow(lastKnownRevision);
+        final AtomicReference<ReadLockStamp> readLockStamp = new AtomicReference<>();
+        return CompletableFuture.runAsync(() -> {
+            readLockStamp.set(readLock(null));
+        }, repositoryWorker).thenCompose(unused -> {
+            return findLatestRevision(ctx, readLockStamp.get(), "watch", normLastKnownRevision, pathPattern);
+        }).thenCompose(latestRevision -> {
+            if (latestRevision != null) {
+                return CompletableFuture.completedFuture(latestRevision);
+            }
 
-            final Revision normLastKnownRevision = normalizeNow(lastKnownRevision);
-            readLock();
             try {
-                // If lastKnownRevision is outdated already and the recent changes match,
-                // there's no need to watch.
-                final Revision latestRevision = blockingFindLatestRevision(normLastKnownRevision, pathPattern);
-                if (latestRevision != null) {
-                    future.complete(latestRevision);
-                } else {
-                    commitWatchers.add(normLastKnownRevision, pathPattern, future);
-                }
+                final CompletableFuture<Revision> future = new CompletableFuture<>();
+                commitWatchers.add(normLastKnownRevision, pathPattern, future);
+                return future;
             } finally {
-                readUnlock();
+                readLockStamp.getAndSet(null).unlock();
+            }
+        }).whenComplete((unused1, unused2) -> {
```

Comment:
nit: How about using `handle` instead of `whenComplete` for performance?

Reply:
Switched to `handle()`